### PR TITLE
PROGRAMMES-6130 Add Radio Favourites Button

### PR DESCRIPTION
**Pull request by votemike**
https://jira.dev.bbc.co.uk/browse/PROGRAMMES-6130

Will write tests soon as long as no one completely disagrees with the structure of the PR.

### CODE REVIEW

**Comment by programmes-developers** [comment]
Can one of the admins verify this patch?

File: config/services.yaml [hunk 1]
```diff
@@ -22,6 +22,8 @@ parameters:
     app.ada.class_service.class: App\ExternalApi\Ada\Service\AdaClassService
     app.electron.baseurl: 'https://api.live.bbc.co.uk/electron'
     app.electron.service.class: App\ExternalApi\Electron\Service\ElectronService
+    app.favourites.url: 'https://www.bbc.co.uk/modules/plugin/musicfavourite/init'
```
Comment (votemike) [inline config/services.yaml]: Do we want to point to LIVE by default?

File: src/Controller/FindByPid/TlecController.php [hunk 2]
```diff
@@ -134,6 +136,11 @@ public function __invoke(
 
         $this->setIstatsLabelsForTlec($onDemandEpisode, $upcomingBroadcast, $lastOn);
 
+        $favouritesButton = null;
+        if ($programme->isRadio()) {
```
Comment (votemike) [inline src/Controller/FindByPid/TlecController.php]: I'm only checking if the programme is a radio programme here but V2 checks all this: https://github.com/bbc/programmes/blob/develop/webapp/php/sharedmodules/programmes/views/helpers/HasFavouritesButton.php#L15
Do we really need all that? Or is this sufficient?

File: src/DsAmen/Presenters/Section/Map/SubPresenter/programme_info.html.twig [hunk 3]
```diff
@@ -10,5 +10,6 @@
                 {{ ds_amen('synopsis', programme_info.getProgramme().getSynopses(), 250) }}
             </div>
         {% endif %}
+        <div class="favourites-button buttons--favourites-small"></div>
```
Comment (votemike) [inline src/DsAmen/Presenters/Section/Map/SubPresenter/programme_info.html.twig]: Acceptable class name?

File: src/DsAmen/Presenters/Section/Map/_map.scss [hunk 4]
```diff
@@ -79,6 +79,17 @@ $map-gap: 2px;
     margin-right: 11px; // due to weird paddings etc
 }
 
+.buttons--favourites-small {
```
Comment (votemike) [inline src/DsAmen/Presenters/Section/Map/_map.scss]: OK for these 2 new styles to live here?

File: config/services_prod_int.yaml [hunk 5]
```diff
@@ -2,4 +2,5 @@ parameters:
     app.feed_env.orbit: 'test'
     app.feed_env.branding: 'live'
     app.electron.baseurl: 'https://api.test.bbc.co.uk/electron'
+    app.favourites.url: 'https://www.test.bbc.co.uk/modules/plugin/musicfavourite/init'
```
Comment (BjornTwachtmann) [inline config/services_prod_int.yaml]: Is there an int version? Does it matter what env the HTML comes from (i.e. will this work on int with the HTML from test?)

File: src/Controller/FindByPid/TlecController.php (hunk 2 above)
Comment (BjornTwachtmann) [reply src/Controller/FindByPid/TlecController.php]: This will do I think

File: src/DsAmen/Presenters/Section/Map/_map.scss (hunk 4 above)
Comment (BjornTwachtmann) [reply src/DsAmen/Presenters/Section/Map/_map.scss]: The favourites button is used elsewhere, might be an idea to put it somewhere more generic. Maybe a file under Css\Components

File: src/ExternalApi/FavouritesButton/FavouritesButton.php [hunk 6]
```diff
@@ -0,0 +1,34 @@
+<?php
+
+namespace App\ExternalApi\FavouritesButton;
```
Comment (BjornTwachtmann) [inline src/ExternalApi/FavouritesButton/FavouritesButton.php]: If you look at the other ExternalApi bits they have a folder structure for Domain , Service etc. Would be good to keep that consistent

File: config/services.yaml (hunk 1 above)
Comment (felipeparaujo) [reply config/services.yaml]: If it works, I think it's fine to point to live

File: src/ExternalApi/FavouritesButton/FavouritesButtonClient.php [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+declare(strict_types = 1);
+
+namespace App\ExternalApi\FavouritesButton;
+
+use App\ExternalApi\Client\HttpApiClientFactory;
+use App\ExternalApi\Exception\ParseException;
+use Closure;
+use GuzzleHttp\Psr7\Response;
+
+class FavouritesButtonClient
```
Comment (BjornTwachtmann) [inline src/ExternalApi/FavouritesButton/FavouritesButtonClient.php]: The other external API fetchy things are called SomethingService

File: templates/find_by_pid/tlec.html.twig [hunk 8]
```diff
@@ -11,6 +11,8 @@
     {% if programme.getNetwork() and programme.getNetwork().isInternational() %}
         {% do preload(asset('js/timezone-notification.js'), { as: 'script', nopush: true }) %}
     {% endif %}
+    {{ favouritesButton.getHead()|raw }}
+    <script>{{ favouritesButton.getScript()|raw }}</script>
```
Comment (BjornTwachtmann) [inline templates/find_by_pid/tlec.html.twig]: Should this go in the head or the foot? Foot seems less likely to block things, but I don't know what they put in the JS.

File: config/services_prod_int.yaml (hunk 5 above)
Comment (votemike) [reply config/services_prod_int.yaml]: There is an int version. But orbit and electron are pointing at test and branding is pointing at live in this file.... so not sure what the right thing to do is....
The different envs result in different URLs in the retrieved HTML.

File: src/ExternalApi/FavouritesButton/FavouritesButtonClient.php (hunk 7 above)
Comment (votemike) [reply src/ExternalApi/FavouritesButton/FavouritesButtonClient.php]: I was basing this off of Orbit, which calls it Client. I can change it to service if you like.

File: templates/find_by_pid/tlec.html.twig (hunk 8 above)
Comment (votemike) [reply templates/find_by_pid/tlec.html.twig]: Could probably go in either.

File: src/Controller/FindByPid/TlecController.php (hunk 2 above)
Comment (felipeparaujo) [reply src/Controller/FindByPid/TlecController.php]: ` $service->product_group->has_favourites` this worries me a bit though... How is this field getting populated?

File: config/services_prod_int.yaml (hunk 5 above)
Comment (BjornTwachtmann) [reply config/services_prod_int.yaml]: yeah, I'm not sure test orbit on int is really ideal either. Change this one to int for now and we'll see how it goes.

File: src/Controller/FindByPid/TlecController.php (hunk 2 above)
Comment (BjornTwachtmann) [reply src/Controller/FindByPid/TlecController.php]: as far as I can see it's hardcoded for radio stuff here https://github.com/bbc/programmes/blob/develop/webapp/php/sharedlib/BBC/Programmes/ProductGroup.php#L33

File: config/services.yaml (hunk 1 above)
Comment (BPScott) [reply config/services.yaml]: name it `app.favourites.baseurl` to be consistent with the naming of other ExternalApi params?

**Comment by BjornTwachtmann** [comment]
LGTM if you want to merge

**Pull request merged**